src/lando/utils/github_checks.py

```diff
@@ -0,0 +1,285 @@
+import logging
+from abc import ABC, abstractmethod
+
+from django.http import HttpRequest
+from typing_extensions import override
+
+from lando.main.models.repo import Repo
+from lando.utils.github import GitHubAPIClient, PullRequest
+from lando.utils.landing_checks import Check
+
+logger = logging.getLogger("__name__")
+
+
+class PullRequestCheck(Check, ABC):
+    @classmethod
+    @abstractmethod
+    def run(
+        cls,
+        pull_request: PullRequest,
+        target_repo: Repo,
+        request: HttpRequest,
+    ) -> list[str]:
+        """Inspect the PR for on issue, and return a message string if present."""
+
+
+#
+# BLOCKERS
+#
+
+
+class PullRequestBlocker(PullRequestCheck, ABC):
+    """Parent class for blocker checks."""
+
+
+class PullRequestUserSCMLevelBlocker(PullRequestBlocker):
+    """You have insufficient permissions to land or your access has expired."""
+
+    @override
+    @classmethod
+    def name(cls) -> str:
+        return "PullRequestUserSCMLevelBlocker"
+
+    @override
+    @classmethod
+    def description(cls) -> str:
+        return "You have insufficient permissions to land or your access has expired."
+
+    @override
+    @classmethod
+    def run(
+        cls,
+        pull_request: PullRequest,
+        target_repo: Repo,
+        request: HttpRequest,
+    ) -> list[str]:
+        # We specifically check the direct user permissions, rather than the union of
+        # those that could have been inherited from group or other roles (e.g., admin).
+        if target_repo.required_permission in request.user.get_user_permissions():
+            return []
+
+        return [cls.description()]
+
+
+# XXX: Irrelevant.
+# class PullRequestUnsupportedRepoBlocker(PullRequestBlocker):
+#     """Repository is not supported by Lando."""
+#
+#     @override
+#     @classmethod
+#     def run(cls, client: GitHubAPIClient, pull_request: PullRequest, target_repo: Repo, request: HttpRequest, request: HttpRequest) -> list[str]:
+#         raise NotImplementedError
+
+
+# XXX: Not currently needed.
+# class PullRequestOpenParentsBlocker(PullRequestBlocker):
+#     """Depends on multiple open parents."""
+#
+#     @override
+#     @classmethod
+#     def run(cls, client: GitHubAPIClient, pull_request: PullRequest, target_repo: Repo, request: HttpRequest, request: HttpRequest) -> list[str]:
+#         raise NotImplementedError
+
+
+class PullRequestClosedBlocker(PullRequestBlocker):
+    """Revision is closed."""
+
+    @override
+    @classmethod
+    def name(cls) -> str:
+        return "PullRequestClosedBlocker"
+
+    @override
+    @classmethod
+    def description(cls) -> str:
+        return "Revision is closed."
+
+    @override
+    @classmethod
+    def run(
+        cls,
+        pull_request: PullRequest,
+        target_repo: Repo,
+        request: HttpRequest,
+    ) -> list[str]:
+        if pull_request.state == pull_request.State.CLOSED:
+            return [cls.description()]
```

Comment:
This is out of scope, but I think in the future it would be better if the checks system returns a state for each check, rather than the warning/error string, and the string can be fetched at rendering time.

Reply:
I think so, yeah. https://bugzilla.mozilla.org/show_bug.cgi?id=2001013